File: tests/wasm/test_ehlers_ecema.js
/**
 * WASM binding tests for EHLERS_ECEMA indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import fs from 'fs';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let wasmInst; 
let testData;
let HAS_WASM_MEMORY = false;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
        
        const wasmBgPath = path.join(path.dirname(importPath.replace('file:///', '/')), 'vector_ta_bg.wasm');
        const wasmBytes = fs.readFileSync(wasmBgPath);
        if (typeof wasm.initSync === 'function') {
            wasmInst = wasm.initSync(wasmBytes);
        } else if (typeof wasm.default === 'function') {
            
            wasmInst = await wasm.default(wasmBytes);
        }

        
        HAS_WASM_MEMORY = !!(wasmInst && wasmInst.memory);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('EHLERS_ECEMA partial params', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.ehlers_ecema_js(close, 20, 50);
    assert.strictEqual(result.length, close.length);
});

test('EHLERS_ECEMA accuracy', () => {
    
    const expected = EXPECTED_OUTPUTS.ehlersEcema;
    
    const data = new Float64Array(testData.close);
    
    
    const length = expected.defaultParams.length;
    const gainLimit = expected.defaultParams.gainLimit;
    
    
    const result = wasm.ehlers_ecema_js(data, length, gainLimit);
    
    assert.strictEqual(result.length, data.length);
    
    
    for (let i = 0; i < expected.warmupPeriod; i++) {
        assert(isNaN(result[i]), `Expected NaN during warmup at index ${i}`);
    }
    
    
    assert(!isNaN(result[expected.warmupPeriod]), "Expected valid value after warmup");
    
    
    const last5 = result.slice(-5);
    assertArrayClose(
        last5,
        expected.last5Values,
        1e-8,
        "EHLERS_ECEMA last 5 values mismatch"
    );
});

test('EHLERS_ECEMA Pine mode accuracy', () => {
    if (!HAS_WASM_MEMORY) {
        console.log('Skipping Pine mode zero-copy test: wasm memory export unavailable');
        return;
    }
    
    const expected = EXPECTED_OUTPUTS.ehlersEcema;
    
    const data = new Float64Array(testData.close);
    
    
    if (wasm.ehlers_ecema_into_ex) {
        const len = data.length;
        const ptr = wasm.ehlers_ecema_alloc(len);
        
        
        const memory = wasmInst.memory;
        const memView = new Float64Array(memory.buffer, ptr, len);
        memView.set(data);
        
        
        const outPtr = wasm.ehlers_ecema_alloc(len);
        
        
        wasm.ehlers_ecema_into_ex(ptr, outPtr, len, 20, 50, true, false);
        
        
        const result = new Float64Array(wasmInst.memory.buffer, outPtr, len);
        const resultCopy = Array.from(result);
        
        
        wasm.ehlers_ecema_free(ptr, len);
        wasm.ehlers_ecema_free(outPtr, len);
        
        
        const last5 = resultCopy.slice(-5);
        assertArrayClose(
            last5,
            expected.pineModeLast5,
            1e-8,
            "EHLERS_ECEMA Pine mode last 5 values mismatch"
        );
        
        
        assert(!isNaN(resultCopy[0]), "Pine mode should have valid value at index 0");
    }
});

test('EHLERS_ECEMA default candles', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.ehlers_ecema_js(close, 20, 50);
    assert.strictEqual(result.length, close.length);
});

test('EHLERS_ECEMA zero period', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.ehlers_ecema_js(inputData, 0, 50);
    }, /Invalid/);
});

test('EHLERS_ECEMA zero gain limit', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.ehlers_ecema_js(inputData, 2, 0);
    }, /Invalid gain limit/);
});

test('EHLERS_ECEMA period exceeds length', () => {
    
    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);
    
    assert.throws(() => {
        wasm.ehlers_ecema_js(dataSmall, 10, 50);
    }, /Invalid/);
});

test('EHLERS_ECEMA very small dataset', () => {
    
    const singlePoint = new Float64Array([42.0]);
    
    assert.throws(() => {
        wasm.ehlers_ecema_js(singlePoint, 20, 50);
    }, /Invalid|Not enough/);
});

test('EHLERS_ECEMA empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.ehlers_ecema_js(empty, 20, 50);
    }, /empty|Empty/);
});

test('EHLERS_ECEMA all NaN input', () => {
    
    const allNan = new Float64Array([NaN, NaN, NaN, NaN, NaN]);
    
    assert.throws(() => {
        wasm.ehlers_ecema_js(allNan, 2, 50);
    }, /All values are NaN/);
});

test('EHLERS_ECEMA invalid gain limit', () => {
    
    const inputData = new Float64Array([10.0, 20.0, 30.0, 40.0, 50.0]);
    
    
    assert.throws(() => {
        wasm.ehlers_ecema_js(inputData, 3, 0);
    }, /Invalid gain limit/);
    
    
    
    
});

test('EHLERS_ECEMA reinput', () => {
    
    const expected = EXPECTED_OUTPUTS.ehlersEcema;
    
    const data = new Float64Array(testData.close);
    
    
    const length = expected.reinputParams.length;
    const gainLimit = expected.reinputParams.gainLimit;
    
    
    const firstResult = wasm.ehlers_ecema_js(data, length, gainLimit);
    assert.strictEqual(firstResult.length, data.length);
    
    
    const secondResult = wasm.ehlers_ecema_js(firstResult, length, gainLimit);
    assert.strictEqual(secondResult.length, firstResult.length);
    
    
    const warmupPeriod = length - 1;
    for (let i = 0; i < warmupPeriod; i++) {
        assert(isNaN(firstResult[i]), `First pass should have NaN in warmup at index ${i}`);
    }
    
    
    
    
    const secondWarmup = warmupPeriod + warmupPeriod;
    for (let i = 0; i < secondWarmup; i++) {
        assert(isNaN(secondResult[i]), `Second pass should have NaN in extended warmup at index ${i}`);
    }
    
    
    assert(!isNaN(firstResult[warmupPeriod]), "First pass should have valid values after warmup");
    
    const validIndices = secondResult.reduce((acc, val, idx) => {
        if (!isNaN(val)) acc.push(idx);
        return acc;
    }, []);
    assert(validIndices.length > 0, "Second pass should have some valid values");
    
    
    const last5 = secondResult.slice(-5);
    assertArrayClose(
        last5,
        expected.reinputLast5,
        1e-8,
        "EHLERS_ECEMA re-input last 5 values mismatch"
    );
});

test('EHLERS_ECEMA NaN handling', () => {
    
    const close = new Float64Array(testData.close);
    
    const result = wasm.ehlers_ecema_js(close, 20, 50);
    assert.strictEqual(result.length, close.length);
    
    
    const warmupPeriod = 19; 
    for (let i = 0; i < warmupPeriod; i++) {
        assert(isNaN(result[i]), `Expected NaN in warmup period at index ${i}`);
    }
    
    
    if (result.length > 240) {
        for (let i = 240; i < result.length; i++) {
            assert(!isNaN(result[i]), `Found unexpected NaN after warmup period at index ${i}`);
        }
    }
    
    
    if (result.length > warmupPeriod) {
        assert(!isNaN(result[warmupPeriod]), `Expected valid value at index ${warmupPeriod}`);
    }
});

test('EHLERS_ECEMA memory management', () => {
    if (!HAS_WASM_MEMORY) {
        console.log('Skipping memory management test: wasm memory export unavailable');
        return;
    }
    
    const expected = EXPECTED_OUTPUTS.ehlersEcema;
    
    const data = new Float64Array(testData.close.slice(0, 25));
    
    const len = data.length;
    const ptr = wasm.ehlers_ecema_alloc(len);
    assert(ptr !== 0, 'Failed to allocate memory');
    
    
    const memView = new Float64Array(wasmInst.memory.buffer, ptr, len);
    memView.set(data);
    
    
    const outPtr = wasm.ehlers_ecema_alloc(len);
    
    
    wasm.ehlers_ecema_into(ptr, outPtr, len, 20, 50);
    
    
    const result = new Float64Array(wasmInst.memory.buffer, outPtr, len);
    const resultCopy = Array.from(result);
    
    
    wasm.ehlers_ecema_free(ptr, len);
    wasm.ehlers_ecema_free(outPtr, len);
    
    
    const regularResult = wasm.ehlers_ecema_js(data, 20, 50);
    assertArrayClose(resultCopy, regularResult, 1e-10, "Memory management test failed");
});

test('EHLERS_ECEMA in-place computation', () => {
    if (!HAS_WASM_MEMORY) {
        console.log('Skipping in-place computation test: wasm memory export unavailable');
        return;
    }
    
    const expected = EXPECTED_OUTPUTS.ehlersEcema;
    
    const data = new Float64Array(testData.close.slice(0, 25));
    
    const len = data.length;
    const ptr = wasm.ehlers_ecema_alloc(len);
    
    
    const memView = new Float64Array(wasmInst.memory.buffer, ptr, len);
    memView.set(data);
    
    
    wasm.ehlers_ecema_into(ptr, ptr, len, 20, 50);
    
    
    const result = new Float64Array(wasmInst.memory.buffer, ptr, len);
    const resultCopy = Array.from(result);
    
    
    wasm.ehlers_ecema_free(ptr, len);
    
    
    const regularResult = wasm.ehlers_ecema_js(data, 20, 50);
    assertArrayClose(resultCopy, regularResult, 1e-10, "In-place computation test failed");
});


test('EHLERS_ECEMA batch single parameter set', () => {
    
    if (!wasm.ehlers_ecema_batch) {
        console.log('EHLERS_ECEMA batch API not yet available');
        return;
    }
    
    const expected = EXPECTED_OUTPUTS.ehlersEcema;
    const close = new Float64Array(testData.close);
    
    
    const batchResult = wasm.ehlers_ecema_batch(close, {
        length_range: [expected.defaultParams.length, expected.defaultParams.length, 0],
        gain_limit_range: [expected.defaultParams.gainLimit, expected.defaultParams.gainLimit, 0]
    });
    
    
    const singleResult = wasm.ehlers_ecema_js(close, expected.defaultParams.length, expected.defaultParams.gainLimit);
    
    assert.strictEqual(batchResult.values.length, singleResult.length);
    assertArrayClose(batchResult.values, singleResult, 1e-10, "Batch vs single mismatch");
    
    
    assert.strictEqual(batchResult.rows, 1);
    assert.strictEqual(batchResult.cols, close.length);
    assert.strictEqual(batchResult.combos.length, 1);
    assert.strictEqual(batchResult.combos[0].length, expected.defaultParams.length);
    assert.strictEqual(batchResult.combos[0].gain_limit, expected.defaultParams.gainLimit);
});

test('EHLERS_ECEMA batch multiple parameters', () => {
    
    if (!wasm.ehlers_ecema_batch) {
        console.log('EHLERS_ECEMA batch API not yet available');
        return;
    }
    
    const expected = EXPECTED_OUTPUTS.ehlersEcema;
    const close = new Float64Array(testData.close.slice(0, 100)); 
    
    
    const batchResult = wasm.ehlers_ecema_batch(close, {
        length_range: expected.batchParams.lengthRange,
        gain_limit_range: expected.batchParams.gainLimitRange
    });
    
    
    assert.strictEqual(batchResult.rows, expected.batchCombinations);
    assert.strictEqual(batchResult.cols, 100);
    assert.strictEqual(batchResult.values.length, expected.batchCombinations * 100);
    
    
    assert.strictEqual(batchResult.combos.length, expected.batchCombinations);
    
    
    const lengths = [15, 20, 25];
    const gains = [40, 50, 60];
    let rowIdx = 0;
    
    for (const length of lengths) {
        for (const gain of gains) {
            const rowStart = rowIdx * 100;
            const rowEnd = rowStart + 100;
            const rowData = batchResult.values.slice(rowStart, rowEnd);
            
            const singleResult = wasm.ehlers_ecema_js(close, length, gain);
            assertArrayClose(
                rowData,
                singleResult,
                1e-10,
                `Length ${length}, gain ${gain} mismatch`
            );
            rowIdx++;
        }
    }
});


test('EHLERS_ECEMA_into_ex with mode flags', () => {
    if (!HAS_WASM_MEMORY) {
        console.log('Skipping into_ex flags test: wasm memory export unavailable');
        return;
    }
    
    if (!wasm.ehlers_ecema_into_ex) {
        console.log('ehlers_ecema_into_ex not yet available');
        return;
    }
    
    
    const data = new Float64Array(testData.close.slice(0, 25));
    
    const len = data.length;
    
    
    const testCases = [
        { pine: false, confirmed: false, desc: "Regular mode" },
        { pine: true, confirmed: false, desc: "Pine mode" },
        { pine: false, confirmed: true, desc: "Confirmed mode" },
        { pine: true, confirmed: true, desc: "Pine + Confirmed mode" }
    ];
    
    for (const testCase of testCases) {
        const inPtr = wasm.ehlers_ecema_alloc(len);
        const outPtr = wasm.ehlers_ecema_alloc(len);
        
        
        const inMem = new Float64Array(wasmInst.memory.buffer, inPtr, len);
        inMem.set(data);
        
        
        wasm.ehlers_ecema_into_ex(
            inPtr, outPtr, len, 20, 50,
            testCase.pine, testCase.confirmed
        );
        
        
        const outMem = new Float64Array(wasmInst.memory.buffer, outPtr, len);
        const result = Array.from(outMem);
        
        
        wasm.ehlers_ecema_free(inPtr, len);
        wasm.ehlers_ecema_free(outPtr, len);
        
        
        assert.strictEqual(result.length, len, `${testCase.desc}: Length mismatch`);
        
        
        if (testCase.pine) {
            
            assert(!isNaN(result[0]), `${testCase.desc}: Should have value at index 0`);
        } else {
            
            for (let i = 0; i < 19; i++) {
                assert(isNaN(result[i]), `${testCase.desc}: Expected NaN at index ${i}`);
            }
            assert(!isNaN(result[19]), `${testCase.desc}: Expected value at index 19`);
        }
    }
});


test('EHLERS_ECEMA zero-copy memory management', () => {
    if (!HAS_WASM_MEMORY) {
        console.log('Skipping zero-copy memory test: wasm memory export unavailable');
        return;
    }
    
    const sizes = [100, 1000, 10000];
    
    for (const size of sizes) {
        const ptr = wasm.ehlers_ecema_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);
        
        
        const memView = new Float64Array(wasmInst.memory.buffer, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 2.5;
        }
        
        
        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 2.5, `Memory corruption at index ${i}`);
        }
        
        
        wasm.ehlers_ecema_free(ptr, size);
    }
});


test('EHLERS_ECEMA SIMD128 consistency', () => {
    
    const testCases = [
        { size: 25, length: 10, gain: 30 },
        { size: 100, length: 20, gain: 50 },
        { size: 1000, length: 30, gain: 70 }
    ];
    
    for (const testCase of testCases) {
        const data = new Float64Array(testCase.size);
        for (let i = 0; i < testCase.size; i++) {
            data[i] = Math.sin(i * 0.1) * 100 + 50000;
        }
        
        const result = wasm.ehlers_ecema_js(data, testCase.length, testCase.gain);
        
        
        assert.strictEqual(result.length, data.length);
        
        
        for (let i = 0; i < testCase.length - 1; i++) {
            assert(isNaN(result[i]), `Expected NaN at warmup index ${i} for size=${testCase.size}`);
        }
        
        
        let sumAfterWarmup = 0;
        let countAfterWarmup = 0;
        for (let i = testCase.length - 1; i < result.length; i++) {
            assert(!isNaN(result[i]), `Unexpected NaN at index ${i} for size=${testCase.size}`);
            sumAfterWarmup += result[i];
            countAfterWarmup++;
        }
        
        
        const avgAfterWarmup = sumAfterWarmup / countAfterWarmup;
        assert(avgAfterWarmup > 40000 && avgAfterWarmup < 60000, 
               `Average value ${avgAfterWarmup} seems unreasonable`);
    }
});

File: tests/python/test_ehlers_ecema.py
"""Python binding tests for EHLERS_ECEMA indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestEhlersEcema:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_ehlers_ecema_partial_params(self, test_data):
        """Test EHLERS_ECEMA with partial parameters (None values) - mirrors test_ehlers_ecema_partial_params"""
        close = test_data['close']


        result = ta_indicators.ehlers_ecema(close, 20, 50)
        assert len(result) == len(close)

    def test_ehlers_ecema_accuracy(self, test_data):
        """Test EHLERS_ECEMA matches expected values from Rust tests - mirrors check_ehlers_ecema_accuracy"""
        expected = EXPECTED_OUTPUTS['ehlers_ecema']

        data = test_data['close']


        length = expected['default_params']['length']
        gain_limit = expected['default_params']['gain_limit']


        result = ta_indicators.ehlers_ecema(data, length, gain_limit)

        assert len(result) == len(data)


        assert np.all(np.isnan(result[:expected['warmup_period']])), "Expected NaN during warmup period"


        assert not np.isnan(result[expected['warmup_period']]), "Expected valid value after warmup"


        assert_close(
            result[-5:],
            expected['last_5_values'],
            rtol=1e-8,
            msg="EHLERS_ECEMA last 5 values mismatch"
        )


        compare_with_rust('ehlers_ecema', result, 'close', expected['default_params'])

    def test_ehlers_ecema_pine_accuracy(self, test_data):
        """Test EHLERS_ECEMA Pine mode matches expected values"""
        expected = EXPECTED_OUTPUTS['ehlers_ecema']

        data = test_data['close']


        try:
            result = ta_indicators.ehlers_ecema(data, 20, 50, pine_compatible=True)


            assert_close(
                result[-5:],
                expected['pine_mode_last_5'],
                rtol=1e-8,
                msg="EHLERS_ECEMA Pine mode last 5 values mismatch"
            )


            assert not np.isnan(result[0]), "Pine mode should have valid value at index 0"


            assert not np.any(np.isnan(result)), "Pine mode should not have any NaN values"
        except TypeError:

            pytest.skip("Pine mode not yet available in Python bindings")

    def test_ehlers_ecema_default_candles(self, test_data):
        """Test EHLERS_ECEMA with default parameters - mirrors test_ehlers_ecema_default_candles"""
        close = test_data['close']


        result = ta_indicators.ehlers_ecema(close, 20, 50)
        assert len(result) == len(close)

    def test_ehlers_ecema_zero_period(self):
        """Test EHLERS_ECEMA fails with zero period - mirrors test_ehlers_ecema_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.ehlers_ecema(input_data, 0, 50)

    def test_ehlers_ecema_zero_gain_limit(self):
        """Test EHLERS_ECEMA fails with zero gain limit"""
        input_data = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid gain limit"):
            ta_indicators.ehlers_ecema(input_data, 2, 0)

    def test_ehlers_ecema_period_exceeds_length(self):
        """Test EHLERS_ECEMA fails when period exceeds data length - mirrors test_ehlers_ecema_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.ehlers_ecema(data_small, 10, 50)

    def test_ehlers_ecema_very_small_dataset(self):
        """Test EHLERS_ECEMA fails with insufficient data - mirrors test_ehlers_ecema_very_small_dataset"""
        single_point = np.array([42.0])

        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.ehlers_ecema(single_point, 20, 50)

    def test_ehlers_ecema_empty_input(self):
        """Test EHLERS_ECEMA fails with empty input - mirrors test_ehlers_ecema_empty_input"""
        empty = np.array([])

        with pytest.raises(ValueError, match="Input data slice is empty|Empty input data"):
            ta_indicators.ehlers_ecema(empty, 20, 50)

    def test_ehlers_ecema_all_nan_input(self):
        """Test EHLERS_ECEMA fails with all NaN input - mirrors test_ehlers_ecema_all_nan"""
        all_nan = np.array([np.nan, np.nan, np.nan, np.nan, np.nan])

        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.ehlers_ecema(all_nan, 2, 50)

    def test_ehlers_ecema_invalid_gain_limit(self):
        """Test EHLERS_ECEMA fails with invalid gain limit - mirrors check_ehlers_ecema_invalid_gain_limit"""
        input_data = np.array([10.0, 20.0, 30.0, 40.0, 50.0])


        with pytest.raises(OverflowError, match="can't convert negative int to unsigned"):
            ta_indicators.ehlers_ecema(input_data, 3, -10)

    def test_ehlers_ecema_reinput(self, test_data):
        """Test EHLERS_ECEMA applied twice (re-input) - mirrors check_ehlers_ecema_reinput"""
        expected = EXPECTED_OUTPUTS['ehlers_ecema']

        data = test_data['close']


        length = expected['reinput_params']['length']
        gain_limit = expected['reinput_params']['gain_limit']


        first_result = ta_indicators.ehlers_ecema(data, length, gain_limit)
        assert len(first_result) == len(data)


        second_result = ta_indicators.ehlers_ecema(first_result, length, gain_limit)
        assert len(second_result) == len(first_result)


        warmup_period = length - 1
        assert np.all(np.isnan(first_result[:warmup_period])), "First pass should have NaN in warmup"




        second_warmup = warmup_period + warmup_period
        assert np.all(np.isnan(second_result[:second_warmup])), "Second pass should have extended warmup"


        assert not np.isnan(first_result[warmup_period]), "First pass should have valid values after warmup"

        valid_indices = np.where(~np.isnan(second_result))[0]
        assert len(valid_indices) > 0, "Second pass should have some valid values"


        assert_close(
            second_result[-5:],
            expected['reinput_last_5'],
            rtol=1e-8,
            msg="EHLERS_ECEMA re-input last 5 values mismatch"
        )

    def test_ehlers_ecema_nan_handling(self, test_data):
        """Test EHLERS_ECEMA handles NaN values correctly - mirrors check_ehlers_ecema_nan_handling"""
        close = test_data['close']

        result = ta_indicators.ehlers_ecema(close, 20, 50)
        assert len(result) == len(close)


        warmup_period = 19
        assert np.all(np.isnan(result[:warmup_period])), f"Expected NaN in warmup period (first {warmup_period} values)"


        if len(result) > 240:
            assert not np.any(np.isnan(result[240:])), "Found unexpected NaN after warmup period"


        if len(result) > warmup_period:
            assert not np.isnan(result[warmup_period]), f"Expected valid value at index {warmup_period}"

    def test_ehlers_ecema_batch_processing(self, test_data):
        """Test EHLERS_ECEMA batch processing - mirrors check_batch_default_row"""
        expected = EXPECTED_OUTPUTS['ehlers_ecema']
        close = test_data['close']


        batch_params = expected['batch_params']
        result = ta_indicators.ehlers_ecema_batch(
            close,
            batch_params['length_range'],
            batch_params['gain_limit_range']
        )


        assert result['rows'] == expected['batch_combinations']
        assert result['cols'] == len(close)

        assert result['values'].shape == (expected['batch_combinations'], len(close))


        assert 'lengths' in result, "Missing lengths array in batch result"
        assert 'gain_limits' in result, "Missing gain_limits array in batch result"
        assert len(result['lengths']) == expected['batch_combinations']
        assert len(result['gain_limits']) == expected['batch_combinations']



        test_params = [
            (15, 40),
            (20, 50),
            (25, 60),
        ]

        for length, gain_limit in test_params:

            single_result = ta_indicators.ehlers_ecema(close, length, gain_limit)



            length_idx = (length - 15) // 5
            gain_idx = (gain_limit - 40) // 10
            row_idx = length_idx * 3 + gain_idx


            assert result['lengths'][row_idx] == length, f"Length mismatch at row {row_idx}"
            assert result['gain_limits'][row_idx] == gain_limit, f"Gain limit mismatch at row {row_idx}"


            batch_row = result['values'][row_idx]



            single_valid_mask = ~np.isnan(single_result)
            first_valid_idx = np.where(single_valid_mask)[0][0] if np.any(single_valid_mask) else 0


            assert_close(
                batch_row[first_valid_idx:],
                single_result[first_valid_idx:],
                rtol=1e-10,
                msg=f"Batch vs single mismatch for length={length}, gain_limit={gain_limit} (from first valid index)"
            )

    def test_ehlers_ecema_batch_default_row(self, test_data):
        """Test EHLERS_ECEMA batch with default parameters"""
        expected = EXPECTED_OUTPUTS['ehlers_ecema']
        close = test_data['close']


        default_params = expected['default_params']
        result = ta_indicators.ehlers_ecema_batch(
            close,
            (default_params['length'], default_params['length'], 0),
            (default_params['gain_limit'], default_params['gain_limit'], 0)
        )


        assert result['rows'] == 1
        assert result['cols'] == len(close)

        assert result['values'].shape == (1, len(close))


        assert result['lengths'][0] == default_params['length']
        assert result['gain_limits'][0] == default_params['gain_limit']


        single_result = ta_indicators.ehlers_ecema(
            close,
            default_params['length'],
            default_params['gain_limit']
        )

        batch_row = result['values'][0]



        single_valid_mask = ~np.isnan(single_result)


        first_valid_idx = np.where(single_valid_mask)[0][0] if np.any(single_valid_mask) else 0


        assert_close(
            batch_row[first_valid_idx:],
            single_result[first_valid_idx:],
            rtol=1e-10,
            msg="Batch default row doesn't match single calculation (from first valid index)"
        )

    def test_ehlers_ecema_stream(self, test_data):
        """Test EHLERS_ECEMA streaming functionality - mirrors check_ehlers_ecema_streaming"""
        expected = EXPECTED_OUTPUTS['ehlers_ecema']

        data = test_data['close']


        stream = ta_indicators.EhlersEcemaStream(
            expected['default_params']['length'],
            expected['default_params']['gain_limit']
        )

        stream_results = []
        for value in data:

            try:
                result = stream.update(value)
                stream_results.append(result if result is not None else np.nan)
            except AttributeError:

                result = stream.next(value)
                stream_results.append(result)


        batch_result = ta_indicators.ehlers_ecema(
            np.array(data),
            expected['default_params']['length'],
            expected['default_params']['gain_limit']
        )


        warmup_period = expected['warmup_period']
        for i in range(warmup_period):
            assert np.isnan(stream_results[i]), f"Expected NaN during warmup at index {i}"
            assert np.isnan(batch_result[i]), f"Expected batch NaN during warmup at index {i}"



        for i in range(warmup_period, len(data)):
            assert_close(
                stream_results[i],
                batch_result[i],
                rtol=1e-6,
                atol=2.0,
                msg=f"Stream vs batch mismatch at index {i}"
            )


        stream.reset()
        try:
            first_value = stream.update(data[0])
            assert first_value is None, "First value after reset should be None (NaN)"
        except AttributeError:
            first_value = stream.next(data[0])
            assert np.isnan(first_value), "First value after reset should be NaN"


        for i in range(1, warmup_period):
            try:
                val = stream.update(data[i])
                assert val is None, f"Expected None during warmup after reset at index {i}"
            except AttributeError:
                val = stream.next(data[i])
                assert np.isnan(val), f"Expected NaN during warmup after reset at index {i}"


        try:
            val = stream.update(data[warmup_period])
            assert val is not None, "Expected valid value after warmup following reset"
        except AttributeError:
            val = stream.next(data[warmup_period])
            assert not np.isnan(val), "Expected valid value after warmup following reset"


if __name__ == "__main__":
    pytest.main([__file__, "-v"])